Add Check Support for Java 21 Unnamed Variables & Patterns Syntax: SuppressWarnings

child https://github.com/checkstyle/checkstyle/issues/14942
I have read check documentation: https://checkstyle.org/checks/annotation/suppresswarnings.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

---


> Allows to specify what warnings that @SuppressWarnings is not allowed to suppress. You can also specify a list of TokenTypes that the configured warning(s) cannot be suppressed on.

I can use @SuppressWarnings on pattern variables. should the check have `PATTERN_VARIABLE_DEF` in the tokens

---
```
PS D:\CS\test> cat src/Test.java                                                
record ColoredPoint(int p, int x, int c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }

public class Test {
    void test(Object obj) {
        if (obj instanceof ColoredPoint( @SuppressWarnings("") int _, _,_)) { } // should this be violation ?
        if (obj instanceof @SuppressWarnings("") Rectangle r) { } // should this be violation ?
        @SuppressWarnings("") int _ = 0; // violation
    }
}

PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="SuppressWarnings">
        </module>
    </module>
</module>
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:8:27: The warning '' cannot be suppressed at this location. [SuppressWarnings]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 

```

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/SuppressWarningsCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/annotation/SuppressWarningsCheck.java
@@ -102,7 +102,9 @@ import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#COMPACT_CTOR_DEF">
  * COMPACT_CTOR_DEF</a>,
  * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_DEF">
- * RECORD_DEF</a>.
+ * RECORD_DEF</a>,
+ * <a href="https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#PATTERN_VARIABLE_DEF">
+ * PATTERN_VARIABLE_DEF</a>.
  * </li>
  * </ul>
  * <p>
@@ -176,6 +178,7 @@ public class SuppressWarningsCheck extends AbstractCheck {
             TokenTypes.CTOR_DEF,
             TokenTypes.COMPACT_CTOR_DEF,
             TokenTypes.RECORD_DEF,
+            TokenTypes.PATTERN_VARIABLE_DEF,
         };
     }
 

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/annotation/SuppressWarningsCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/annotation/SuppressWarningsCheck.xml
@@ -45,7 +45,7 @@
                <description>Specify the RegExp to match against warnings. Any warning
  being suppressed matching this pattern will be flagged.</description>
             </property>
-            <property default-value="CLASS_DEF,INTERFACE_DEF,ENUM_DEF,ANNOTATION_DEF,ANNOTATION_FIELD_DEF,ENUM_CONSTANT_DEF,PARAMETER_DEF,VARIABLE_DEF,METHOD_DEF,CTOR_DEF,COMPACT_CTOR_DEF,RECORD_DEF"
+            <property default-value="CLASS_DEF,INTERFACE_DEF,ENUM_DEF,ANNOTATION_DEF,ANNOTATION_FIELD_DEF,ENUM_CONSTANT_DEF,PARAMETER_DEF,VARIABLE_DEF,METHOD_DEF,CTOR_DEF,COMPACT_CTOR_DEF,RECORD_DEF,PATTERN_VARIABLE_DEF"
                       name="tokens"
                       type="java.lang.String[]"
                       validation-type="tokenSet">

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/annotation/SuppressWarningsCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/annotation/SuppressWarningsCheckTest.java
@@ -1081,4 +1081,19 @@ public class SuppressWarningsCheckTest extends AbstractModuleTestSupport {
                 getNonCompilablePath("InputSuppressWarningsRecords.java"), expected);
     }
 
+    @Test
+    public void testSuppressWarningsPatternVariables() throws Exception {
+
+        final String[] expected = {
+            "18:57: " + getCheckMessage(MSG_KEY_SUPPRESSED_WARNING_NOT_ALLOWED, ""),
+            "21:62: " + getCheckMessage(MSG_KEY_SUPPRESSED_WARNING_NOT_ALLOWED, "unused"),
+            "24:44: " + getCheckMessage(MSG_KEY_SUPPRESSED_WARNING_NOT_ALLOWED, "unchecked"),
+            "27:27: " + getCheckMessage(MSG_KEY_SUPPRESSED_WARNING_NOT_ALLOWED, ""),
+        };
+
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputSuppressWarningsPatternVariables.java"),
+                expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/annotation/suppresswarnings/InputSuppressWarningsPatternVariables.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/annotation/suppresswarnings/InputSuppressWarningsPatternVariables.java
@@ -1,0 +1,40 @@
+/*
+SuppressWarnings
+format = ^unchecked$|^unused$|^$
+tokens = (default)CLASS_DEF, INTERFACE_DEF, ENUM_DEF, ANNOTATION_DEF, ANNOTATION_FIELD_DEF, \
+         ENUM_CONSTANT_DEF, PARAMETER_DEF, VARIABLE_DEF, METHOD_DEF, CTOR_DEF, \
+         COMPACT_CTOR_DEF, RECORD_DEF, PATTERN_VARIABLE_DEF
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.annotation.suppresswarnings;
+
+public class InputSuppressWarningsPatternVariables {
+
+    void test(Object o) {
+        // violation below, 'The warning '' cannot be suppressed at this location.'
+        if (o instanceof ColoredPoint(@SuppressWarnings("")int x, int y, int z)) { }
+
+        // violation below, 'The warning 'unused' cannot be suppressed at this location.'
+        if (o instanceof ColoredPoint(_, _,@SuppressWarnings("unused") int _)) { }
+
+        // violation below, 'The warning 'unchecked' cannot be suppressed at this location.'
+        if (o instanceof @SuppressWarnings("unchecked") ColoredPoint _) { }
+
+        // violation below, 'The warning '' cannot be suppressed at this location.'
+        @SuppressWarnings("")
+        int _ = sideEffect();
+
+        if (o instanceof @SuppressWarnings("a unused") ColoredPoint _) { }
+
+        @SuppressWarnings("foo")
+        int _ = sideEffect();
+    }
+
+    int sideEffect() {
+        return 0;
+    }
+    record ColoredPoint(int x, int y, int z) { }
+}

--- a/src/xdocs/checks/annotation/suppresswarnings.xml
+++ b/src/xdocs/checks/annotation/suppresswarnings.xml
@@ -92,6 +92,8 @@
                     COMPACT_CTOR_DEF</a>
                 , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_DEF">
                     RECORD_DEF</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#PATTERN_VARIABLE_DEF">
+                    PATTERN_VARIABLE_DEF</a>
                   .
               </td>
               <td>
@@ -119,6 +121,8 @@
                     COMPACT_CTOR_DEF</a>
                 , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#RECORD_DEF">
                     RECORD_DEF</a>
+                , <a href="../../apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#PATTERN_VARIABLE_DEF">
+                    PATTERN_VARIABLE_DEF</a>
                   .
               </td>
               <td>5.0</td>

